packages/core/src/mesh/ModelMesh.ts

```diff
@@ -14,4 +67,693 @@ export class ModelMesh extends Mesh {
     super(engine);
     this.name = name;
   }
+
+  /**
+   * Upload Mesh Data to the graphics API.
+   * @param noLongerAccessible - Whether to access data later. If true, you'll never access data anymore (free memory cache)
+   */
+  uploadData(noLongerAccessible: boolean): void {
```

Comment:
I write a optimization-version. mianly change is vertex dara update,use the nealy logic with your index data update. This version solve  some  bug、more strict update logic and less code.
```typescript
/**
   * Upload Mesh Data to the graphics API.
   * @param noLongerAccessible - Whether to access data later. If true, you'll never access data anymore (free memory cache)
   */
  uploadData(noLongerAccessible: boolean): void {
    const { _indices, _vertexSlotChanged, _accessible } = this;
    if (!_accessible) {
      throw "Not allowed to access data while accessible is false.";
    }

    // Vertex element change.
    if (_vertexSlotChanged) {
      const vertexElements = this._updateVertexElements();
      this.setVertexElements(vertexElements);
      this._vertexSlotChanged = false;
    }

    // Vertex Value change.
    const vertexBuffer = this._vertexBufferBindings[0]?._buffer;
    const elementCount = this._elementCount;
    let vertices = this._verticesFloat32;
    if (!vertexBuffer || vertices.byteLength !== elementCount * this._vertexCount) {
      vertexBuffer && vertexBuffer.destroy();
      vertices = new Float32Array(elementCount * this._vertexCount);
      this._verticesUint8 = new Uint8Array(vertices.buffer);
      this._verticesFloat32 = vertices;

      this._vertexChangeFlag = ValueChanged.All;
      this._updateVertices(vertices);

      const newVertexBuffer = new Buffer(
        this._engine,
        BufferBindFlag.VertexBuffer,
        vertices,
        noLongerAccessible ? BufferUsage.Static : BufferUsage.Dynamic
      );
      this._setVertexBufferBinding(0, new VertexBufferBinding(newVertexBuffer, elementCount * 4));
    } else if (this._vertexChangeFlag & ValueChanged.All) {
      const vertices = this._verticesFloat32;
      this._updateVertices(vertices);
      vertexBuffer.setData(vertices);
    }

    const indexBuffer = this._indexBufferBinding?._buffer;
    if (_indices) {
      if (!indexBuffer || _indices.byteLength != indexBuffer.byteLength) {
        indexBuffer && indexBuffer.destroy();
        const indexBufferNew = new Buffer(this._engine, BufferBindFlag.IndexBuffer, _indices);
        this._setIndexBufferBinding(new IndexBufferBinding(indexBufferNew, this._indicesFormat));
      } else if (this._indicesChangeFlag) {
        this._indicesChangeFlag = false;
        indexBuffer.setData(_indices);
        if (this._indexBufferBinding._format !== this._indicesFormat) {
          this._setIndexBufferBinding(new IndexBufferBinding(indexBuffer, this._indicesFormat));
        }
      }
    } else if (indexBuffer) {
      indexBuffer.destroy();
      this._setIndexBufferBinding(null);
    }

    if (noLongerAccessible) {
      this._accessible = false;
      this._releaseCache();
    }
  }
```

Reply:
+1